### PR TITLE
Issue-47: Install Recommended Plugins as Part of Install

### PR DESCRIPTION
**Pull request by mogmarsh**
# Pull Request

## Related Issue
[Install Recommended Plugins as Part of Install](https://github.com/alleyinteractive/create-wordpress-project/issues/47)

## Description
This pull request enhances the installation process by automatically installing recommended plugins based on user consent. It installs plugins using composer commands called from PHP. This allows for prompting for user input to install. optional plugins and conditionally installing plugins for Pantheon.

#### Default Required Plugin List
- [Alleyvate](https://github.com/alleyinteractive/wp-alleyvate)
- [Byline Manager](https://github.com/alleyinteractive/byline-manager)
- [Elasticsearch Extensions](https://github.com/alleyinteractive/elasticsearch-extensions)
- [Meta Inspector](https://github.com/alleyinteractive/meta-inspector)
- [MSM Sitemap](https://github.com/Automattic/msm-sitemap)
- [Safe Redirect Manager](https://wordpress.org/plugins/safe-redirect-manager/)
- [WP Asset Manager](https://github.com/alleyinteractive/wp-asset-manager)
- [WP Curate](https://github.com/alleyinteractive/wp-curate)
- [WP New Relic Transactions](https://github.com/alleyinteractive/wp-new-relic-transactions)
- [Yoast SEO](https://wordpress.org/plugins/wordpress-seo/)

#### Suggested Plugins
- [ES Admin](https://github.com/alleyinteractive/es-admin)
- [ES WP Query](https://github.com/alleyinteractive/es-wp-query)
- [Fieldmanager](https://github.com/alleyinteractive/wordpress-fieldmanager)
- [Yoast Duplicate Post](https://wordpress.org/plugins/duplicate-post/)

#### Install Automatically if Pantheon
- [Pantheon Advanced Page Cache](https://github.com/pantheon-systems/pantheon-advanced-page-cache)
- [Query Monitor](https://wordpress.org/plugins/query-monitor/)
- [Rewrite Rules Inspector](https://wordpress.org/plugins/rewrite-rules-inspector/)
- [SearchPress](https://github.com/alleyinteractive/searchpress)
- [SearchPress Debug Bar](https://github.com/alleyinteractive/debug-bar-searchpress)
- [Two Factor](https://wordpress.org/plugins/two-factor/)
- [WP Mail SMTP](https://wordpress.org/plugins/wp-mail-smtp/)
- [WP Native PHP Sessions](https://wordpress.org/plugins/wp-native-php-sessions/)
- [Object Cache Pro](https://objectcache.pro/docs/composer-installation) (requires a token for Composer installation)

Note: The Pantheon MU Plugin is installed by default and removed by the script already. We can decide if we want to switch this over later.

## Use Case
For developers to easily install Alley's recommended list of plugins when creating a new site.

## Changes and Updates
- Plugins can now be installed to an optional custom path.
- The Pantheon MU Plugin was excluded from the install list due to issues.
- An `auth.json` file is now included to facilitate installations requiring authentication, such as Object Cache Pro.
- Fixed the install path for SearchPress.
- Resolved various plugin-related issues and removed redundant code.
- The installation process can execute quietly without excessive output.
- After usage, the `composer-templates` is cleaned up by deletion.
- The requirement for `ext-redis` is now ignored, to avoid conflicts in environments where it isn't available.

## Comments from Discussion
- Proposed by @srtfisher:
  - [Meta Inspector](https://github.com/alleyinteractive/meta-inspector)
  - [Alleyvate](https://github.com/alleyinteractive/wp-alleyvate)

---

Please include any relevant details about the implementation and any changes made to the codebase. Remember to update any documentation if necessary.

### Checklist:
- [ ] I have read the [CONTRIBUTING](CONTRIBUTING.md) document.
- [ ] My code follows the code style of this project.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] I have updated the documentation accordingly.
- [ ] This PR includes changes to the database schema or migrations.
- [ ] I have informed stakeholders of my changes.

### CODE REVIEW

File: configure.php [hunk 1]
```diff
@@ -239,6 +239,33 @@ function delete_files( string|array $paths ): void {
 	}
 }
 
+/**
+ * Install a plugin with composer
+ *
+ * @param array $plugin_data The plugin information.
+ * @param bool $prompt Whether or not to prompt to install.
+ * @return void
```
Comment (renatonascalves) [inline configure.php]:
```suggestion
```

File: configure.php [hunk 2]
```diff
@@ -239,6 +239,33 @@ function delete_files( string|array $paths ): void {
...
+ */
+function install_plugin( $plugin_data, $prompt = false ) {
```
Comment (renatonascalves) [inline configure.php]:
```suggestion
function install_plugin( array $plugin_data, bool $prompt = false ): void {
```

File: configure.php [hunk 3]
```diff
@@ -575,6 +602,54 @@ function delete_files( string|array $paths ): void {
 	echo "Done!\n\n";
 }
 
+write( 'Installing Required Plugins...' );
+$required_file_contents = file_get_contents( 'composer-templates/default.json' );
+$required_plugins       = json_decode( $required_file_contents, true );
+foreach( $required_plugins as $plugin ) {
+	install_plugin( $plugin, false );
```
Comment (renatonascalves) [inline configure.php]:
```suggestion
	install_plugin( $plugin );
```

File: configure.php [hunk 4]
```diff
@@ -575,6 +602,54 @@ function delete_files( string|array $paths ): void {
...
+		replace_in_file( 'auth.json', [ 'object_cache_pro_token' => $license_key ] );
+
+		install_plugin(
+			[
+				'name'      => 'Object Cache Pro',
+				'repo'      => 'https://objectcache.pro/repo/',
+				'repo_type' => 'composer',
+				'path'      => 'rhubarbgroup/object-cache-pro'
+			],
+			false
```
Comment (renatonascalves) [inline configure.php]:
```suggestion
			]
```

**Review by renatonascalves — APPROVED** [review]
🌵

File: composer-templates/default.json [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        "path": "alleyinteractive/wp-curate"
+    },
+    {
+        "name": "WP New Relic Transactions",
+        "repo": "https://github.com/alleyinteractive/wp-new-relic-transactions",
+        "path": "alleyinteractive/wp-new-relic-transactions"
+    },
+    {
+        "name": "Yoast SEO",
+        "path": "yoast/wordpress-seo"
```
Comment (kevinfodness) [inline composer-templates/default.json]: Should install from wpackagist

File: composer-templates/pantheon.json [hunk 6]
```diff
@@ -0,0 +1,38 @@
+[
+    {
+        "name": "Pantheon Advanced Page Cache",
+        "path": "pantheon-systems/pantheon-advanced-page-cache"
```
Comment (kevinfodness) [inline composer-templates/pantheon.json]: wpackagist

File: composer-templates/pantheon.json [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    },
+    {
+        "name": "Query Monitor",
+        "path": "johnbillion/query-monitor"
```
Comment (kevinfodness) [inline composer-templates/pantheon.json]: wpackagist

File: composer-templates/pantheon.json [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    },
+    {
+        "name": "Rewrite Rules Inspector",
+        "repo": "https://github.com/Automattic/rewrite-rules-inspector",
+        "path": "Automattic/rewrite-rules-inspector"
```
Comment (kevinfodness) [inline composer-templates/pantheon.json]: wpackagist, and remove the repo

File: composer-templates/pantheon.json [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        "path": "alleyinteractive/searchpress"
+    },
+    {
+        "name": "SearchPress Debug Bar",
+        "repo": "https://github.com/alleyinteractive/debug-bar-searchpress",
+        "path": "alleyinteractive/debug-bar-searchpress"
+    },
+    {
+        "name": "Two Factor",
+        "path": "georgestephanis/two-factor"
```
Comment (kevinfodness) [inline composer-templates/pantheon.json]: wpackagist

File: composer-templates/suggested.json [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+        "repo": "https://github.com/alleyinteractive/es-wp-query",
+        "path": "alleyinteractive/es-wp-query"
+    },
+    {
+        "name": "Fieldmanager",
+        "path": "alleyinteractive/wordpress-fieldmanager"
+    },
+    {
+        "name": "Yoast Duplicate Post",
+        "path": "yoast/duplicate-post"
```
Comment (kevinfodness) [inline composer-templates/suggested.json]: wpackagist

File: configure.php [hunk 11]
```diff
@@ -575,6 +601,53 @@ function delete_files( string|array $paths ): void {
...
+$required_plugins       = json_decode( $required_file_contents, true );
+foreach( $required_plugins as $plugin ) {
+	install_plugin( $plugin );
+}
+
+write( 'Installing Suggested Plugins...' );
+$suggested_file_contents = file_get_contents( 'composer-templates/suggested.json' );
+$suggested_plugins       = json_decode( $suggested_file_contents, true );
+foreach( $suggested_plugins as $plugin ) {
+	install_plugin( $plugin, true );
```
Comment (kevinfodness) [inline configure.php:615]: I'd rather this be interactive, where for each suggested plugin it prompts the user as to whether they want to install it or not, and only does so if they say yes.

**Review by kevinfodness — APPROVED** [review]
Left some suggestions, otherwise 🍣

File: configure.php (hunk 11 above)
Comment (mogmarsh) [reply configure.php:615]: @kevinfodness it does - that's the way this works. The second param on `install_plugin` is `$prompt` if `true`, it asks and only installs if the answer is yes.

Comment (kevinfodness) [reply configure.php:615]: Great - thanks, missed that detail on my initial review.

**Pull request merged**